cpp/src/neighbors/detail/cagra/cagra_search.cuh

```diff
@@ -108,6 +111,109 @@ void search_main_core(raft::resources const& res,
   }
 }
 
+/**
+ * @brief Performs ANN search using brute force when filter sparsity exceeds a specified threshold.
+ *
+ * This function switches to a brute force search approach to improve recall rate when the
+ * `sample_filter` function filters out a high proportion of samples, resulting in a sparsity level
+ * (proportion of unfiltered samples) exceeding the specified `threshold_to_bf`.
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of database vector indices
+ * @tparam internal_IdxT during search we map IdxT to internal_IdxT, this way we do not need
+ * separate kernels for int/uint.
+ *
+ * @param[in] handle
+ * @param[in] params configure the search
+ * @param[in] idx ivf-pq constructed index
+ * @param[in] queries a device matrix view to a row-major matrix [n_queries, index->dim()]
+ * @param[out] neighbors a device matrix view to the indices of the neighbors in the source dataset
+ * [n_queries, k]
+ * @param[out] distances a device matrix view to the distances to the selected neighbors [n_queries,
+ * k]
+ * @param[in] sample_filter a device filter function that greenlights samples for a given query
+ * @param[in] threshold_to_bf A sparsity threshold; brute force is used when sparsity exceeds this
+ * threshold, in the range [0, 1]
+ *
+ * @return true If the brute force search was applied successfully.
+ * @return false If the brute force search was not applied.
+ */
+template <typename T,
+          typename InternalIdxT,
+          typename CagraSampleFilterT,
+          typename IdxT      = uint32_t,
+          typename DistanceT = float>
+bool search_using_brute_force(
+  raft::resources const& res,
+  search_params& params,
+  const index<T, IdxT>& index,
+  raft::device_matrix_view<const T, int64_t, raft::row_major>& queries,
+  raft::device_matrix_view<InternalIdxT, int64_t, raft::row_major>& neighbors,
+  raft::device_matrix_view<DistanceT, int64_t, raft::row_major>& distances,
+  CagraSampleFilterT& sample_filter,
+  double threshold_to_bf = 0.9)
+{
+  bool is_applied = false;
+  auto n_queries  = queries.extent(0);
+  auto n_dataset  = index.size();
+
+  auto bitset_filter_view = sample_filter.bitset_view_;
+  auto dataset_view       = index.contiguous_dataset();
```

Comment:
Good idea! It looks like the CAGRA always holds a strided dataset, am I right?

Reply:
Either strided or compressed at the moment, check the if branches with dynamic casts in the search_main below.

Reply:
Hi @achirkin , thank you for the suggestion, all changes have been done, except moving the `search_using_brute_force` down to the if branches, the consideration is that I hope the logic can remain clean and not mixed together too much, which may be more readable, also I suppose this part of the code may not be stable and can be changed for more format/data type, and I also want to avoid declare some unused objects(like `graph` `graph_internal`, though they are light-weight). If you still recommend to move down, I will make it :)